### PR TITLE
Adds support for getting the location of the URL

### PR DESCRIPTION
**Pull request by jaysaleh**
This pull request gets the `URL` where the image uploaded is stored and prints it to the console for testing purposes. Further steps include adding the `URL` into the `comment` entity.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -25,13 +25,26 @@
...
+import java.util.List;
+import java.util.Map;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
 /** Servlet that writes messages as a response. */
+// TODO: Remove print statement and store image URL in database
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]:
```suggestion
// TODO: Remove print statement and store image URL in database.
```

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 2]
```diff
@@ -41,7 +54,12 @@
   private static final String NAME = "name";
   private static final String COMMENT_TEXT = "commentText";
   private static final String EMAIL = "email";
-  
+
+  /** Supported image files */
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I wouldn't make comments for private members JavaDocs. See: https://google.github.io/styleguide/javaguide.html#s7.3-javadoc-where-required

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -63,11 +81,55 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     commentEntity.setProperty(COMMENT_TEXT, commentText);
     commentEntity.setProperty(TIME_STAMP, timeStamp);
 
+    // Used for testing, see TODO
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: It is always best to add the TODO where the hack is. So in this case, move your class TODO to here.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -63,11 +81,55 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+    System.out.println(getUploadedFileUrl(request, /* formInputElementName= */ "image"));
+
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     datastore.put(commentEntity);
 
     response.sendRedirect(REDIRECT_URL);
   }
+
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload an image file. */
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: It is better to use Optional as it makes it explicit that nothing might get returned.

Comment (billie-cat) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: https://docs.google.com/document/d/1Dq_mcrOdyfbEB_ovC9tQGui8DLopkES2vwEejckLom0/edit#heading=h.5rm21e874sp9

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -63,11 +81,55 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+  private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL. (dev server)
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: What does the comment (dev server) mean?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -63,11 +81,55 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      return null;
+    }
+
+    // Gets first and only file in form submission
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]:
```suggestion
    // Gets first and only file in form submission.
```

Comment (billie-cat) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Please end your comments with periods. There are a few more which need a period.

**Review by billie-cat — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 1 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: done

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: done

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 5 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: dev server is the local server when testing. Changed from dev to local

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 6 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: done

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 2 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: done

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: fixed

Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: fixed

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 7]
```diff
@@ -64,11 +80,56 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+    // TODO: Remove print statement and store image URL in database.
+    System.out.println(getUploadedFileUrl(request, /* formInputElementName= */ "image").get());
+
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     datastore.put(commentEntity);
 
     response.sendRedirect(REDIRECT_URL);
   }
+
+  /** Returns a URL that points to the uploaded file, or null if the user didn't upload an image file. */
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: This JavaDoc needs to be updated to mention Optional.empty() and not null.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 8]
```diff
@@ -64,11 +80,56 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+    // User submitted form without selecting a file, so we can't get a URL. (live server)
+    BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+    if (blobInfo.getSize() == 0) {
+      blobstoreService.delete(blobKey);
+      return Optional.empty();
+    }
+
+    String fileInfo = blobInfo.getContentType();
+
+    // Return null if file is not a jpg, png or tiff image.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Need to update this comment as well.

**Review by billie-cat — CHANGES_REQUESTED** [review]
Concerning "dev server is the local server when testing. Changed from dev to local". The parentheses comments are not complete sentences and I am not sure what they are trying to indicate. If they are important then they should be made into full sentences or else removed.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 9]
```diff
@@ -64,11 +80,56 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+    if (!fileInfo.equals(JPEG) && !fileInfo.equals(PNG) && !fileInfo.equals(TIFF)) {
+      return Optional.empty();
+
+    }
+
+    // Gets URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobKey);
+
+    // Returns relative path to image.
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: This comment explains what the code is doing. You only need comments when it is not clear what is happening from the code. I believe this comment and potentially others can be omitted.

**Review by billie-cat — CHANGES_REQUESTED** [review]

**Review by f-camp — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 7 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: fixed

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 8 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: fixed

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 9 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: fixed

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 10]
```diff
@@ -63,11 +80,56 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+    System.out.println(getUploadedFileUrl(request, /* formInputElementName= */ "image").get());
+
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     datastore.put(commentEntity);
 
     response.sendRedirect(REDIRECT_URL);
   }
+
+  /** 
+   * Returns a URL that points to the uploaded file, or an empty optional
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]:
```suggestion
   * Returns the URL that points to the uploaded file, or an empty optional
```

Comment (billie-cat) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: This JavaDoc doesn't explain what the uploaded file is? Is it {@code formInputElementName}. It also doesn't explain what the arguments are.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 11]
```diff
@@ -41,7 +53,12 @@
   private static final String NAME = "name";
   private static final String COMMENT_TEXT = "commentText";
   private static final String EMAIL = "email";
-  
+
+  // Supported image files.
+  private static final String JPEG = "image/jpeg";
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:58]: Optional: consider creating: List<String> supportedImageFormats 

This way you do not need to create 1 variable per image format.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 12]
```diff
@@ -63,11 +80,56 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+  /** 
+   * Returns a URL that points to the uploaded file, or an empty optional
+   * if the user didn't upload an image file. 
+   */
+  private Optional<String> getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL. (local server)
```
Comment (billie-cat) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Concerning the comment you left: dev server is the local server when testing. Changed from dev to local. The parentheses comments are not complete sentences and I am not sure what they are trying to indicate. If they are important then they should be made into full sentences or else removed.

**Review by billie-cat — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 10 above)
Comment (jaysaleh) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: fixed

**Pull request merged**